README.md
# JavaEE-projects
[![Welcome downloads](https://img.shields.io/badge/downloads-349m%20total-brightgreen.svg)](https://github.com/ScarlettRay/JavaEE-projects/archive/master.zip)

>**存放一些自己写的还有从不同开源社区fork下来的JavaEE项目，其中就不乏一些很多企业单位都在用的源码。**
```lua
├──1.在线教育平台

├──2.jfinal cms

├──3.龙果支付系统

├──4.优客服

├──5.J2EE企业级开发解决方案

├──6.CRM

├──7.OA

├──8.秒杀系统
```


jfinal_cms/ChangeLog.md
Change Log 更新说明
------------------------------

## 2017-06-11 jfinal cms4.7.0
1. 去除jflyfox_jfinal和jflyfox_base项目依赖，将代码合并并重构

## 2017-06-01
1. 修复多站点和单站点切换导致的异常。

## 2017-05-26 jfinal cms4.6.0
1. jflyfox_jfinal升级为4.5.0
2. 重构站点逻辑代码，让后台站点配置与前台站点配置分离；
3. delete、insert、update默认不保存日志信息；如果需要可以默认保存可以恢复BaseProjectModel类；
4. 规范SessionUser和SysUser属性，避免相同属性不同写法问题；
5. 解决ueditor和umeditor可能会报重复注册controller错误；
6. 解决IE提交数据可能保存两条数据问题；

## 2017-05-02
1. 将密码由des改为3des加密。
2. 修复登陆cookie异常，导致不能记住密码问题。
3. 加入用户密码重置功能。
4. 加入api秘钥展示功能。

## 2017-04-17
1. 修复IE11重复提交数据问题。

### 2017-04-13
1. 添加jfinal cms api文档

### 2017-04-05 jfinal cms4.5.0
1. 添加文件恢复机制。
2. 加入阿里云OSS备份和恢复处理。
3. 完善文档，细节优化并加入了站点配置和文件备份配置说明。

### 2017-04-04
1. 完成上传备份机制。
2. ueditor加入上传备份及站点机制。
3. 优化上传路径处理机制。

### 2017-03-23
1. 组织机构根目录展示问题修复。

### 2017-03-22    jfinal cms4.4.0
1. API接口全面完善，加入参数加密和认证功能
2. 可以通过配置切换加密和认证功能开启
3. 修复fst导致第三方Oauth2登陆异常问题

### 2017-03-14
1. 加入API用户类型，组织机构树结构调整

### 2017-03-05
1. 后台登陆加入md5传输，防止网络拦截。

### 2017-02-08    jfinal cms4.3.0
1. 缓存加入redis支持
2. 第三方核心包升级：jfinal v3.0、beetl v2.7.11

### 2017-01-22    jfinal cms4.2.0
1. 优化论坛模板数据
2. 模板发布功能优化

### 2017-01-20   
1. 加入论坛模板

### 2017-01-07   
1. 完善博客样式
2. 加入参数配置菜单更新sql
3. 完善权限细节

### 2017-01-01    jfinal cms4.1.5
1. 后台加入新的主题flat。

### 2016-12-31
1. 优化站点配置后台以及拦截逻辑。
2. 将站点设置加入到参数管理中。
3. 默认为单站点，可以方便的通过站点管理切换。

### 2016-12-17    jfinal cms4.1.4
1. 加入系统参数配置模块。
2. api配置迁移到系统参数配置中。
3. 修复登陆后可以跳过菜单限制，加入菜单权限拦截。

### 2016-11-15
1. api接口调用加入了登陆验证以及验证开关。

### 2016-10-03   jfinal cms4.1.3
1. 完善api功能。
2. 加入api接口文档。

### 2016-10-02
1. 加入api功能。

### 2016-09-21   jfinal cms4.1.2
1. 完善初始化数据。
2. 完善blog模板。
3. 修改默认模板为blog。

### 2016-09-20
1. 加入blog模板。

### 2016-08-01  jfinal cms4.1.1
1. 多站点key冲突问题。

### 2016-07-31
1. 组织机构优化，加入编码和上级组织机构。
2. 组织机构加入树形展示。


### 2016-06-23
1. 修复异常数据问题。
2. 修复sql多版本兼容问题。
3. 修复多站点UrlKey一直导致的不可预知bug。


### 2016-06-21  jfinal cms4.1.0
1. jfinal cms 加入project项目模板。

### 2016-06-12
1. jfinal cms项目文档完善。

### 2016-06-10
1. 加入jfinal cms项目文档。

### 2016-06-01
1. 修复项目路径为根目录时，模板加载异常问题。

### 2016-05-29
1. 修复标签查询图片展示异常问题

### 2016-05-11  jfinal cms4.0.0
1. 升级为4.0.0

### 2016-05-04
1. 友情链接加入多站点支持。
2. 用户属性加入站点配置，避免再次登录重复选择站点问题。
3. 用户登录限制优化，加入用户类型和创建站点设置。
4. 加入菜单权限拦截器，能够阻止大部分异常操作。
5. 支持目录模板路径设置。

### 2016-04-27
1. 修改上传路径~支持多站点目录结构。

### 2016-04-25
1. 多站点和单站点支持优化。

### 2016-04-18
1. 加入ue编辑器
2. 上传目录修改
3. 高亮功能
4. oauth跳转问题修复

### 2016-04-15
1. 站点管理持续优化

### 2016-03-19  jfinal cms3.1.0
1. 升级为3.1.0
2. 修复cookie异常，持续报错问题。
3. 加入文章直接跳转到第三方功能。
4. 实现前后台登陆隔离。
5. 后台菜单引用路径重构。
6. 加入文章审核功能，可以配置是否需要审核。
7. 加入文章复制功能。

### 2016-03-13
1. 加入栏目和文章树结构展示

### 2016-03-06  jfinal cms3.0.0
1. 支持模板文件在线编辑，升级为3.0.0

### 2016-03-06
1. 优化模板管理
2. 加入主题切换。
3. 加入修改文件备份处理。

### 2016-03-04
1. 加入模板管理模块，引入文件管理。
2. filemanager来自github，再此基础上进行了路径优化，主题切换支持，中文支持，以及replaceFile，addFile接口完善。

### 2016-02-24
1. jflyfox_base修改为3.1修复中文和空格目录问题
2. 解决pom中mvn tomcat:run 目录问题

### 2016-02-20 jfinal cms2.9.0
1. 重构为maven项目，升级为2.9.0

### 2016-02-20 jfinal cms2.8.0
1. 升级为2.8.0

### 2016-02-20
1. 完善视频模板。
2. 完善视频模板数据。

### 2016-02-19
1. 加入后台视频专辑和视频模块。
2. 加入前端视频模板。
3. 重构前端图片模板路径。

### 2016-02-14 jfinal cms2.7.0
1. 升级2.7.0。

### 2016-02-10
1. 完善相册管理，加入我的相册。
2. 加入相册标签编辑。
3. 加入相册管理前台模板。
4. 图片查看页面优化。
5. 相册管理模板加入分页展示。

### 2016-02-5
1. 加入相册管理和图片管理。

### 2016-02-03
1. 加入后台首页面。
2. 登录成功后，跳转到首页面。

### 2016-01-31 jfinal cms2.6.0
1. 加入前台栏目公告。
2. 优化

### 2016-01-29
1. 意见反馈优化
2. 优化菜单选择，重新整理菜单

### 2016-01-28
1. 加入意见反馈，栏目滚屏，栏目公告三个模块后台代码。
2. 迁移前台首页滚屏数据。
3. 加入前台意见反馈。

### 2016-01-28 jfinal cms2.5.0
1. 升级为2.5.0。
12. 后台加入主题切换，bootstrap默认，黑色，flat-ui三个主题。

### 2016-01-20 jfinal cms2.4.0
1. 升级为2.4.0。
2. 项目前台页面全面重构，jfinal升级为2.2。
3. 前台加入模板扩展功能，通过beetl functions获取后台数据。
原有jfinal cms模板不变，加入website官网模板。

### 2016-01-19
1. jfinal升级为2.2。

### 2016-01-16
1. 升级为2.3.0。
2. 项目全面重构，jfinal升级为2.1。
3. 后台管理支持列表排序查询功能。
4. 加入项目基类model和controller解决和依赖包基类重复代码覆盖问题。

### 2016-01-15
1. 名字更新，升级为2.2.0

### 2016-01-14 mtgxxw2.1.1
1. 升级父类依赖，升级为2.1.1

### 2016-01-03 mtgxxw2.1.0
1. 代码全面重构，优化目录结构，升级为2.1
2. 加入日志管理，记录后台增删改查以及登陆登出操作。
3. 完善前台目录url美化。

### 2015-12-27
1. 修复评论分页显示bug

### 2015-11-29 mtgxxw2.0.0
1. 主要讲jfinal升级为2.0

### 2015-09-26
1. 加入文章附件功能：确保存在download目录（tmp临时目录，image_url图片目录，file_url附件目录）
在文章修改目录进行上传和删除操作，附件只支持一个文件，现在是可以上传所有类型。

### 2015-09-21
1. 实现文章缓存自动更新，解决文章发布不可以见问题（需要后台点击更新缓存）。
config.properties文件中可以配置CMS.UPDATE_TIME字段更新时间，单位毫秒；配置0为不更新。
2. 修复用户发布文章中更新时间为空问题。

### 2015-08-16 mtgxxw1.4.2
1. 加入喜欢的文章。
2. 解决cookie访问异常，导致页面无法访问问题。
3. 修改我的空间页面布局。

### 2015-08-04
1. 感谢老李提醒，mysql5.1中group_concat可能到返回类型是blob而不能转换为String，mysql5.5没有这个问题。
可以通过CAST(group_concat(tagname) AS CHAR)进行优化，未进行测试。

### 2015-08-03
1. 修复cookie获取权限后，导致后台菜单加载失败问题。

### 2015-08-02 mtgxxw1.4.1
1. 加入用户cookie，避免重启服务用户信息失效以及关闭浏览器失效问题
2. 修改登陆页面与页面风格保持一致，加入验证码。
3. 修复我的空间页面bug。
4. 修复首页跳转bug。

### 2015-06-21 mtgxxw1.4
1. 用户登陆后可以进行博文发布、编辑、删除，发布加入字符限制。
2. Ueditor图片上传加入大小和数量限制。
3. 加入非法上传用户记录。

### 2015-06-13 mtgxxw1.3
1. 优化首页，加入最新新闻和最新动态切换。
2. 优化头像点击展示。

### 2015-06-09 mtgxxw1.2
1. 发布1.2版本，主要是bug修复。
2. 感谢fxhover同学，修复漏洞：http://www.jflyfox.com/mtg/front/tags/%3Cimg%20%20src=%22xx%22%20onerror=%22alert%28111%29;%22%3E111

### 2015-06-06
1. 加入点击用户查看信息。

### 2015-06-04
1. jflyfox_jfinal升级为2.1,全面去除模板，更简单易懂
2. 登录名称错误修改。
3. 优化文章编辑操作。

### 2015-06-03 mtgxxw1.1
1. 发布1.1版本，这次主要优化前台访问速度。
2. 加入cdn，如果本地没有网络请修改/static/component/include/jquery.html和/static/component/include/bootstrap.html。

### 2015-06-02
1. 回复管理加入排序。
2. 前台查看加入伪静态化。
3. 优化用户部门，加入第三方注册部门。
4. 根据加载慢问题，加入外链图片加载，缓解前台带宽问题。首页加入网络图片，后台加入编辑。

### 2015-05-31 mtgxxw1.0
1. 基本功能页面已完成，可以进行开源。
2. 如还包含私人数据，希望大家提醒我进行删除。

### 2015-05-27
1. 完善首页和菜单seo。
2. 优化model，加入get set方法。
3. 加入评论数和浏览量计数缓存。
4. 完善关于我们。
5. 功能已基本完善，待完善数据。


### 2015-05-26
1. 已完成初稿，基本可以使用。
2. 加入菜单url优化，简洁美观。
3. 加入伪静态化，看着更高大上。

### 2015-05-25
1. 初始化，完成首页


ukefu(优客服)/src/main/java/com/ukefu/util/MD5.java
package com.ukefu.util;

public class MD5 {

	public String getMD5ofStr(String inbuf) {
		md5Init();
		md5Update(inbuf.getBytes(), inbuf.length());
		md5Final();
		digestHexStr = "";
		for (int i = 0; i < 16; i++)
			digestHexStr += byteHEX(digest[i]);

		return digestHexStr;
	}
	
	public String getMD5ofByte(byte[] inbuf) {
		md5Init();
		md5Update(inbuf, inbuf.length);
		md5Final();
		digestHexStr = "";
		for (int i = 0; i < 16; i++)
			digestHexStr += byteHEX(digest[i]);

		return digestHexStr;
	}
	
	public MD5() {
		state = new long[4];
		count = new long[2];
		buffer = new byte[64];
		digest = new byte[16];
		md5Init();
	}

	private void md5Init() {
		count[0] = 0L;
		count[1] = 0L;
		state[0] = 0x67452301L;
		state[1] = 0xefcdab89L;
		state[2] = 0x98badcfeL;
		state[3] = 0x10325476L;
	}

	private long F(long x, long y, long z) {
		return x & y | ~x & z;
	}

	private long G(long x, long y, long z) {
		return x & z | y & ~z;
	}

	private long H(long x, long y, long z) {
		return x ^ y ^ z;
	}

	private long I(long x, long y, long z) {
		return y ^ (x | ~z);
	}

	private long FF(long a, long b, long c, long d, long x, long s, long ac) {
		a += F(b, c, d) + x + ac;
		a = (int) a << (int) s | (int) a >>> (int) (32L - s);
		a += b;
		return a;
	}

	private long GG(long a, long b, long c, long d, long x, long s, long ac) {
		a += G(b, c, d) + x + ac;
		a = (int) a << (int) s | (int) a >>> (int) (32L - s);
		a += b;
		return a;
	}

	private long HH(long a, long b, long c, long d, long x, long s, long ac) {
		a += H(b, c, d) + x + ac;
		a = (int) a << (int) s | (int) a >>> (int) (32L - s);
		a += b;
		return a;
	}

	private long II(long a, long b, long c, long d, long x, long s, long ac) {
		a += I(b, c, d) + x + ac;
		a = (int) a << (int) s | (int) a >>> (int) (32L - s);
		a += b;
		return a;
	}

	private void md5Update(byte inbuf[], int inputLen) {
		byte block[] = new byte[64];
		int index = (int) (count[0] >>> 3) & 0x3f;
		if ((count[0] += inputLen << 3) < (long) (inputLen << 3))
			count[1]++;
		count[1] += inputLen >>> 29;
		int partLen = 64 - index;
		int i;
		if (inputLen >= partLen) {
			md5Memcpy(buffer, inbuf, index, 0, partLen);
			md5Transform(buffer);
			for (i = partLen; i + 63 < inputLen; i += 64) {
				md5Memcpy(block, inbuf, 0, i, 64);
				md5Transform(block);
			}

			index = 0;
		} else {
			i = 0;
		}
		md5Memcpy(buffer, inbuf, index, i, inputLen - i);
	}

	private void md5Final() {
		byte bits[] = new byte[8];
		encode(bits, count, 8);
		int index = (int) (count[0] >>> 3) & 0x3f;
		int padLen = index >= 56 ? 120 - index : 56 - index;
		md5Update(PADDING, padLen);
		md5Update(bits, 8);
		encode(digest, state, 16);
	}

	private void md5Memcpy(byte output[], byte input[], int outpos, int inpos, int len) {
		for (int i = 0; i < len; i++)
			output[outpos + i] = input[inpos + i];

	}

	private void md5Transform(byte block[]) {
		long a = state[0];
		long b = state[1];
		long c = state[2];
		long d = state[3];
		long x[] = new long[16];
		decode(x, block, 64);
		a = FF(a, b, c, d, x[0], 7L, 0xd76aa478L);
		d = FF(d, a, b, c, x[1], 12L, 0xe8c7b756L);
		c = FF(c, d, a, b, x[2], 17L, 0x242070dbL);
		b = FF(b, c, d, a, x[3], 22L, 0xc1bdceeeL);
		a = FF(a, b, c, d, x[4], 7L, 0xf57c0fafL);
		d = FF(d, a, b, c, x[5], 12L, 0x4787c62aL);
		c = FF(c, d, a, b, x[6], 17L, 0xa8304613L);
		b = FF(b, c, d, a, x[7], 22L, 0xfd469501L);
		a = FF(a, b, c, d, x[8], 7L, 0x698098d8L);
		d = FF(d, a, b, c, x[9], 12L, 0x8b44f7afL);
		c = FF(c, d, a, b, x[10], 17L, 0xffff5bb1L);
		b = FF(b, c, d, a, x[11], 22L, 0x895cd7beL);
		a = FF(a, b, c, d, x[12], 7L, 0x6b901122L);
		d = FF(d, a, b, c, x[13], 12L, 0xfd987193L);
		c = FF(c, d, a, b, x[14], 17L, 0xa679438eL);
		b = FF(b, c, d, a, x[15], 22L, 0x49b40821L);
		a = GG(a, b, c, d, x[1], 5L, 0xf61e2562L);
		d = GG(d, a, b, c, x[6], 9L, 0xc040b340L);
		c = GG(c, d, a, b, x[11], 14L, 0x265e5a51L);
		b = GG(b, c, d, a, x[0], 20L, 0xe9b6c7aaL);
		a = GG(a, b, c, d, x[5], 5L, 0xd62f105dL);
		d = GG(d, a, b, c, x[10], 9L, 0x2441453L);
		c = GG(c, d, a, b, x[15], 14L, 0xd8a1e681L);
		b = GG(b, c, d, a, x[4], 20L, 0xe7d3fbc8L);
		a = GG(a, b, c, d, x[9], 5L, 0x21e1cde6L);
		d = GG(d, a, b, c, x[14], 9L, 0xc33707d6L);
		c = GG(c, d, a, b, x[3], 14L, 0xf4d50d87L);
		b = GG(b, c, d, a, x[8], 20L, 0x455a14edL);
		a = GG(a, b, c, d, x[13], 5L, 0xa9e3e905L);
		d = GG(d, a, b, c, x[2], 9L, 0xfcefa3f8L);
		c = GG(c, d, a, b, x[7], 14L, 0x676f02d9L);
		b = GG(b, c, d, a, x[12], 20L, 0x8d2a4c8aL);
		a = HH(a, b, c, d, x[5], 4L, 0xfffa3942L);
		d = HH(d, a, b, c, x[8], 11L, 0x8771f681L);
		c = HH(c, d, a, b, x[11], 16L, 0x6d9d6122L);
		b = HH(b, c, d, a, x[14], 23L, 0xfde5380cL);
		a = HH(a, b, c, d, x[1], 4L, 0xa4beea44L);
		d = HH(d, a, b, c, x[4], 11L, 0x4bdecfa9L);
		c = HH(c, d, a, b, x[7], 16L, 0xf6bb4b60L);
		b = HH(b, c, d, a, x[10], 23L, 0xbebfbc70L);
		a = HH(a, b, c, d, x[13], 4L, 0x289b7ec6L);
		d = HH(d, a, b, c, x[0], 11L, 0xeaa127faL);
		c = HH(c, d, a, b, x[3], 16L, 0xd4ef3085L);
		b = HH(b, c, d, a, x[6], 23L, 0x4881d05L);
		a = HH(a, b, c, d, x[9], 4L, 0xd9d4d039L);
		d = HH(d, a, b, c, x[12], 11L, 0xe6db99e5L);
		c = HH(c, d, a, b, x[15], 16L, 0x1fa27cf8L);
		b = HH(b, c, d, a, x[2], 23L, 0xc4ac5665L);
		a = II(a, b, c, d, x[0], 6L, 0xf4292244L);
		d = II(d, a, b, c, x[7], 10L, 0x432aff97L);
		c = II(c, d, a, b, x[14], 15L, 0xab9423a7L);
		b = II(b, c, d, a, x[5], 21L, 0xfc93a039L);
		a = II(a, b, c, d, x[12], 6L, 0x655b59c3L);
		d = II(d, a, b, c, x[3], 10L, 0x8f0ccc92L);
		c = II(c, d, a, b, x[10], 15L, 0xffeff47dL);
		b = II(b, c, d, a, x[1], 21L, 0x85845dd1L);
		a = II(a, b, c, d, x[8], 6L, 0x6fa87e4fL);
		d = II(d, a, b, c, x[15], 10L, 0xfe2ce6e0L);
		c = II(c, d, a, b, x[6], 15L, 0xa3014314L);
		b = II(b, c, d, a, x[13], 21L, 0x4e0811a1L);
		a = II(a, b, c, d, x[4], 6L, 0xf7537e82L);
		d = II(d, a, b, c, x[11], 10L, 0xbd3af235L);
		c = II(c, d, a, b, x[2], 15L, 0x2ad7d2bbL);
		b = II(b, c, d, a, x[9], 21L, 0xeb86d391L);
		state[0] += a;
		state[1] += b;
		state[2] += c;
		state[3] += d;
	}

	private void encode(byte output[], long input[], int len) {
		int i = 0;
		for (int j = 0; j < len; j += 4) {
			output[j] = (byte) (int) (input[i] & 255L);
			output[j + 1] = (byte) (int) (input[i] >>> 8 & 255L);
			output[j + 2] = (byte) (int) (input[i] >>> 16 & 255L);
			output[j + 3] = (byte) (int) (input[i] >>> 24 & 255L);
			i++;
		}

	}

	private void decode(long output[], byte input[], int len) {
		int i = 0;
		for (int j = 0; j < len; j += 4) {
			output[i] = b2iu(input[j]) | b2iu(input[j + 1]) << 8 | b2iu(input[j + 2]) << 16 | b2iu(input[j + 3]) << 24;
			i++;
		}

	}

	private long b2iu(byte b) {
		return b >= 0 ? b : b & 0xff;
	}

	private String byteHEX(byte ib) {
		char Digit[] = { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f' };
		char ob[] = new char[2];
		ob[0] = Digit[ib >>> 4 & 0xf];
		ob[1] = Digit[ib & 0xf];
		String s = new String(ob);
		return s;
	}

	static final int S11 = 7;
	static final int S12 = 12;
	static final int S13 = 17;
	static final int S14 = 22;
	static final int S21 = 5;
	static final int S22 = 9;
	static final int S23 = 14;
	static final int S24 = 20;
	static final int S31 = 4;
	static final int S32 = 11;
	static final int S33 = 16;
	static final int S34 = 23;
	static final int S41 = 6;
	static final int S42 = 10;
	static final int S43 = 15;
	static final int S44 = 21;
	static final byte PADDING[] = { -128, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0 };
	private long state[];
	private long count[];
	private byte buffer[];
	private String digestHexStr;
	private byte digest[];

}